### PR TITLE
Multi threaded bruteforce

### PR DESCRIPTION
**Pull request by ArtemiyGranat**
I think I will rename `my_sem_X` to `sem_X`, move it to `semaphore.h` file and add `include` based on OS if my semaphore implementation is correct.

Something like:
```
#ifdef __APPLE__
#include "semaphore.h"
#else
#include <semaphore.h>
#endif
```

### CODE REVIEW

File: main.c [hunk 1]
```diff
@@ -12,13 +28,96 @@ typedef enum
...
+  pthread_cond_init (&sem->cond_sem, NULL);
+  pthread_mutex_init (&sem->mutex, NULL);
+}
+
+void
+my_sem_post (my_sem_t *sem)
+{
+  pthread_mutex_lock (&sem->mutex);
+  ++sem->counter;
+  pthread_cond_signal (&sem->cond_sem);
```
Comment (hnarkaytis) [inline main.c:64]: Signalling is an expensive kernel call. It make sense to make it only if counter was 0 before increment.

File: main.c [hunk 2]
```diff
@@ -12,13 +28,96 @@ typedef enum
...
+  pthread_mutex_unlock (&sem->mutex);
+}
+
+void
```
Comment (hnarkaytis) [inline main.c:68]: Error code

File: main.c [hunk 3]
```diff
@@ -12,13 +28,96 @@ typedef enum
...
+  queue->head = queue->tail = 0;
+
+  my_sem_init (&queue->full, 0, 0);
+  my_sem_init (&queue->empty, 0, QUEUE_SIZE);
+
+  pthread_mutex_init (&queue->head_mutex, NULL);
+  pthread_mutex_init (&queue->tail_mutex, NULL);
+}
+
+void
```
Comment (hnarkaytis) [inline main.c:100]: Error code

File: main.c [hunk 4]
```diff
@@ -12,13 +28,96 @@ typedef enum
...
+{
+  my_sem_wait (&queue->empty);
+  pthread_mutex_lock (&queue->tail_mutex);
+  queue->queue[queue->tail] = *task;
+  queue->tail = (queue->tail + 1) % QUEUE_SIZE;
+  pthread_mutex_unlock (&queue->tail_mutex);
+  my_sem_post (&queue->full);
+}
+
+void
```
[the remark this answers is not in the document]
Comment (ArtemiyGranat) [reply main.c:111]: Should I add messages to errors in `queue_X` functions?
Like `fprintf (stderr, "Could not initialize semaphore");`

File: main.c [hunk 5]
```diff
@@ -12,13 +28,170 @@ typedef enum
...
+    }
+  while (sem->counter == 0)
+    {
+      if (pthread_cond_wait (&sem->cond_sem, &sem->mutex) != 0)
+        {
+          return -1;
+        }
+    }
+  --sem->counter;
+  pthread_mutex_unlock (&sem->mutex);
```
Comment (hnarkaytis) [inline main.c:107]: error check?

**Review by hnarkaytis — CHANGES_REQUESTED** [review]
One more minor error check and this could be merged.

**Review by hnarkaytis — APPROVED** [review]
LGTM

**Pull request merged**